crawler/crawler-ftp/src/main/java/fr/pilato/elasticsearch/crawler/fs/crawler/ftp/FileAbstractorFTP.java

```diff
@@ -0,0 +1,215 @@
+        } else {
+            fullPath = new String(fullPath.getBytes(ALTERNATIVE_ENCODING), FTP.DEFAULT_CONTROL_ENCODING);
+        }
+
+        InputStream inputStream = ftp.retrieveFileStream(fullPath);
+        if (inputStream != null) {
+            return inputStream;
+        } else {
+            throw new IOException(String.format("FTP client can not retrieve stream for [%s]", file.getFullpath()));
+        }
+    }
+
+    @Override
+    public void closeInputStream(InputStream inputStream) throws IOException {
+        inputStream.close();
+        // This is necessary if we want to read the stream
+        ftp.completePendingCommand();
+    }
+
+    @Override
+    public Collection<FileAbstractModel> getFiles(String dir) throws IOException {
+        logger.debug("Listing files from {}", dir);
+        if (isUtf8) {
+            dir = new String(dir.getBytes(StandardCharsets.UTF_8), FTP.DEFAULT_CONTROL_ENCODING);
+        } else {
+            dir = new String(dir.getBytes(ALTERNATIVE_ENCODING), FTP.DEFAULT_CONTROL_ENCODING);
+        }
+
+        FTPFile[] ftpFiles = ftp.listFiles(dir);
+        if (ftpFiles == null) return null;
+        List<FTPFile> files = Arrays.stream(ftpFiles).filter(file -> {
+            if (fsSettings.getFs().isFollowSymlinks()) return true;
+            return !file.isSymbolicLink();
+        }).collect(Collectors.toList());
+
+        Collection<FileAbstractModel> result = new ArrayList<>(files.size());
+        // Iterate other files
+        // We ignore here all files like . and ..
+        String finalDir = dir;
+        result.addAll(files.stream().filter(file -> !".".equals(file.getName()) &&
+                !"..".equals(file.getName()))
+                .map(file -> toFileAbstractModel(finalDir, file))
+                .collect(Collectors.toList()));
+
+        logger.debug("{} local files found", result.size());
+        return result;
+    }
+
+    @Override
+    public boolean exists(String dir) {
+        try {
+            logger.debug("Checking dir existence: " + dir);
+            if (isUtf8) {
+                dir = new String(dir.getBytes(StandardCharsets.UTF_8), FTP.DEFAULT_CONTROL_ENCODING);
+            } else {
+                dir = new String(dir.getBytes(ALTERNATIVE_ENCODING), FTP.DEFAULT_CONTROL_ENCODING);
+            }
+            return ftp.changeWorkingDirectory(dir);
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public void open() throws IOException {
+        Server server = fsSettings.getServer();
+        logger.debug("Opening FTP connection to {}@{}", server.getUsername(), server.getHostname());
+
+        ftp = new FTPClient();
+        ftp.addProtocolCommandListener(ftpListener);
+        // send a safe command (i.e. NOOP) over the control connection to reset the router's idle timer
+        ftp.setControlKeepAliveTimeout(300);
```

Comment:
Wondering if we should in the future make this configurable... (Not needed for this PR)

Reply:
Yes, there are many things could be configurable. By the way, I learned a lot from [FileZilla](https://filezilla-project.org/) logs. It's a good reference.